tao_compiler/mlir/disc/transforms/disc_comp_intens_fusion_to_cuda_source.cc

```diff
@@ -0,0 +1,631 @@
+    return false;
+  }
+  SmallVector<std::string> layouts;
+  if (!getLayoutStrings(dot.getDotDimensionNumbers(), layouts)) {
+    return false;
+  }
+  element_a_layout = layouts[0];
+  element_b_layout = layouts[1];
+  element_output_layout = layouts[2];
+
+  if (!getAccumulatorTypeString(d_type.getElementType(),
+                                element_accumulator_type) ||
+      !getOperatorClassTypeString(operator_class_type) ||
+      !getSMArchString(sm_arch) || !getScaleKindString(scale_kind) ||
+      !getCountVectorizedString(element_output_type, count_vectorized) ||
+      !getPermuteDLayoutString(func, permute_d_layout)) {
+    return false;
+  }
+
+  // TODO: support more epilogue types.
+  epilogue_type = element_output_type;
+  gather_a = isGatherA(func) ? "true" : "false";
+  gather_b = isGatherB(func) ? "true" : "false";
+  scatter_d = isScatterD(func) ? "true" : "false";
+
+  std::string intent = "    ";
+  SourceEmitterCUDA source_emitter;
+  SourceEmitterCUDA::ValueNameBinding binding;
+  for (auto& op : func.getRegion().getOps()) {
+    if (isa<lmhlo::DotGeneralOp>(&op)) {
+      // Convert CUTLASS datatype to default cuda datatype for computation.
+      Value dot_output = op.getOperand(2);
+      std::string old_name = "input";
+      std::string new_name;
+      SmallVector<std::string> convert_instructions;
+      if (!mayConvertCutlassTypeToCUDAType(dot_output, old_name, new_name,
+                                           convert_instructions)) {
+        return false;
+      }
+      source_emitter.bindValueNames(SmallVector<Value>({dot_output}),
+                                    SmallVector<std::string>({new_name}),
+                                    binding);
+      for (auto& instruction : convert_instructions) {
+        specialized_epilogue += intent + instruction + ";\n";
+      }
+    } else if (!isa<func::ReturnOp>(&op)) {
+      assert(source_emitter.isSupportedOp(&op) && "Encounter unsupported op.");
+      auto instruction = source_emitter.EmitOp(&op, binding);
+      if (!instruction.hasValue()) {
+        return false;
+      } else {
+        specialized_epilogue += intent + instruction.value() + ";\n";
+      }
+    }
+  }
+  // Append return instruction.
+  SmallVector<Value> results;
+  getResults(func, results);
+  // Only support one result currently.
+  // TODO: support more outputs.
+  if (results.size() != 1) {
+    return false;
+  }
+  Value result = results[0];
+  std::string old_name = binding[result];
+  std::string new_name;
+  SmallVector<std::string> convert_result_instructions;
+  if (!mayConvertCUDATypeToCutlassType(result, old_name, new_name,
+                                       convert_result_instructions)) {
+    return false;
+  }
+  for (auto& instruction : convert_result_instructions) {
+    specialized_epilogue += intent + instruction + ";\n";
+  }
+  specialized_epilogue += intent + "return " + new_name + ";";
```

Comment:
How about creating a thin wrapper to help `specialized_epilogue` to hide useless details like `indent` and `;\n`. Something like:


```c++
Helper h;
h.appendLine(instruction);
h.appendLine(instruction_part1, instruction_part2);
// ....
std::string specialized_eplilogue = h.finalize();
```